app/graphql/types/query_type.rb

```diff
@@ -119,6 +119,40 @@ def search_media_by_title(title:)
     service.search(title)
   end
 
+  field :random_media, Types::Interface::Media, null: false do
+    description 'Random anime or manga'
+    argument :media_type, Types::Enum::MediaType, required: true
+    argument :age_ratings, [Types::Enum::AgeRating],
+      required: true,
+      prepare: ->(age_ratings, context) do
+        # No authorization is needed for sfw shows.
+        return age_ratings if age_ratings.exclude?('R18')
+
+        if context[:token].blank?
+          raise GraphQL::ExecutionError, 'You must be authorized to view R18 media'
+        elsif User.current.sfw_filters?
+          raise GraphQL::ExecutionError, 'You must have SFW filters turned off'
+        end
+
+        age_ratings
+      end
+  end
+
+  def random_media(media_type:, age_ratings:)
+    media = nil
+
+    # Most cases should find a media first time, but there is a chance
+    # due to TABLESAMPLE one won't be found.
+    while media.blank?
+      media = media_type.safe_constantize
+                        .from(Arel.sql("#{media_type.downcase} TABLESAMPLE BERNOULLI(1)"))
```

Comment:
SYSTEM_ROWS(1) for significantly faster table processing for a small reduction in randomness.

Extension needs to be added to psql initialisation (Google psql system rows for extension name)

This can remove the comment above about sometimes not returning a result as it will always return 1 result (table row) as long as the media type (or library) has at least one resource.

Reply:
A migration like this should work to enable the extension:

```ruby
def up
  enable_extension 'tsm_system_rows'
end
```

Reply:
ok, thank you.

Reply:
@wopian @NuckChorris  Just ran this `media_type.safe_constantize.from(Arel.sql("#{media_type.downcase} TABLESAMPLE SYSTEM_ROWS(1)"))` and this literally only returns one record. It still has to filter out the `age_rating` which could result in a null. Not sure what the best way to handle retrying (or if this modification is even worth it?)

Reply:
Don't you want to be filtering before randomly picking anyway? It's still considerably faster than BERNOULLI and the other methods (300μs to 20ms to filter)

Reply:
In fact, with `SELECT * from anime WHERE age_rating = 0 AND random() <= 0.01 LIMIT 1` (1% sample, similar to `BERNOULLI(1)`). Increasing to 10% (`0.1`) has no discernible impact on the query times.

`age_rating` | anime | manga
:-: | :-: | :-:
0 | 80-400μs | 40μs-10ms
1 | 60-400μs | 40μs-900μs
2 | 50-280μs | 80μs-1.2ms
3 | 70-540μs | 230μs-4.2ms

(20x explain analyse each, rounded to 2 significant figures)

Reply:
Sooo... Use what you are saying above?

Reply:
I guess. I didn't realise TABLESAMPLE could only be used on an unfiltered table (you'd need to create a temporary table to do it, which I presume would wipe all/most of the speed benefits it had over the painfully slow initial implementation of `ORDER BY random()`).

Had the requirements been "get _any_ media" then TABLESAMPLE would definitely have been the way to go.